src/Avalonia.Controls/TransitioningContentControl.cs

```diff
@@ -1,33 +1,35 @@
 ﻿using System;
 using System.Threading;
+using System.Threading.Tasks;
 using Avalonia.Animation;
+using Avalonia.Controls.Presenters;
 using Avalonia.Controls.Templates;
-using Avalonia.Threading;
+using Avalonia.Data;
 
 namespace Avalonia.Controls;
 
 /// <summary>
-/// Displays <see cref="ContentControl.Content"/> according to a <see cref="FuncDataTemplate"/>.
-/// Uses <see cref="PageTransition"/> to move between the old and new content values. 
+/// Displays <see cref="ContentControl.Content"/> according to an <see cref="IDataTemplate"/>,
+/// using a <see cref="PageTransition"/> to move between the old and new content. 
 /// </summary>
 public class TransitioningContentControl : ContentControl
 {
-    private CancellationTokenSource? _lastTransitionCts;
-    private object? _currentContent;
+    private CancellationTokenSource? _currentTransition;
+    private ContentPresenter? _transitionPresenter;
+    private Optional<object?> _transitionFrom;
 
     /// <summary>
     /// Defines the <see cref="PageTransition"/> property.
     /// </summary>
     public static readonly StyledProperty<IPageTransition?> PageTransitionProperty =
-        AvaloniaProperty.Register<TransitioningContentControl, IPageTransition?>(nameof(PageTransition),
-            new CrossFade(TimeSpan.FromSeconds(0.125)));
-    
-    /// <summary>
-    /// Defines the <see cref="CurrentContent"/> property.
-    /// </summary>
-    public static readonly DirectProperty<TransitioningContentControl, object?> CurrentContentProperty =
-        AvaloniaProperty.RegisterDirect<TransitioningContentControl, object?>(nameof(CurrentContent),
-            o => o.CurrentContent);
+        AvaloniaProperty.Register<TransitioningContentControl, IPageTransition?>(nameof(PageTransition));
+
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("AvaloniaProperty", "AVP1012",
+        Justification = "Default property values shouldn't be set with SetCurrentValue.")]
+    public TransitioningContentControl()
+    {
+        PageTransition = new CrossFade(TimeSpan.FromMilliseconds(125));
```

Comment:
Not really sure we should be doing this, but the previous version had a default value for `PageTransition`, though it was using a mutable object as a `StyledProperty` default value which means the instance was shared.

We don't want to do that as it means that changing the properties on a `PageTransition` for once `TransitioningContentControl` will mean that all instances with the default value will change.

Would it be an acceptable breaking change to make the default value be `null` and require setting the transition on each `TransitioningContentControl` explicitly?

Reply:
We can keep this property null by default and reuse private crossfade transition instance, when value is unset (not only null, but unset). 

Though, I don't see problems with having no default transition either. In my experience I always had to disable them as annoying anyway, as ReactiveUI routing host is inheriting this control.

Reply:
Can we have an immutable shared default instead? It could proxy calls to the regular CrossFade transition.

Reply:
An immutable shared default might be best. I'd rather not treat null as a crossfade because there may be a valid reason to not want a transition.

Reply:
Done.